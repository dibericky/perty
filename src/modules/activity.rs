#[derive(Debug, PartialEq)]
pub struct Estimation {
    pub optimistic: u32,
    pub probable: u32,
    pub pessimistic: u32,
}

impl Estimation {
    pub fn estimated(&self) -> f64 {
        let tm = f64::from(self.probable);
        let to = f64::from(self.optimistic);
        let tp = f64::from(self.pessimistic);
        1f64 / 6f64 * (4f64 * tm + to + tp)
    }
}

#[derive(Debug, PartialEq)]
pub struct Activity {
    pub name: String,
    pub estimation: Estimation,
}

impl Activity {
    pub fn new(name: String, optimistic: u32, probable: u32, pessimistic: u32) -> Self {
        Self {
            name,
            estimation: Estimation {
                optimistic,
                probable,
                pessimistic,
            },
        }
    }

    pub fn estimated(&self) -> f64 {
        self.estimation.estimated()
    }
}

#[cfg(test)]
mod test {
    use super::Activity;

    #[test]
    fn get_estimated() {
        let estimation = Activity::new("activity 1".to_string(), 6, 10, 15).estimated();
        assert_eq!(estimation, 10.166666666666666);
    }
}
