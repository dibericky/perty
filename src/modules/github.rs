use anyhow::{Context, Result};
use regex::Regex;
use reqwest::header::USER_AGENT;
use serde::Deserialize;

pub type BoardId = i32;

pub struct Github {
    token: String,
}

#[derive(Debug, Deserialize)]
pub struct CreatedBoard {
    pub html_url: String,
    pub id: BoardId,
}

impl Github {
    pub fn new(token: String) -> Self {
        Self { token }
    }

    pub fn create_board(&self, name: String, owner: String, repo: String) -> Result<CreatedBoard> {
        let url = format!("https://api.github.com/repos/{}/{}/projects", owner, repo);
        let response: CreatedBoard = reqwest::blocking::Client::new()
            .post(url)
            .header(USER_AGENT, "perty")
            .bearer_auth(self.token.to_owned())
            .header("content-type", "application/json")
            .json(&serde_json::json!({
                "name": &name,
                "body": "Project created by PERTy: https://github.com/dibericky/perty",
            }))
            .send()?
            .json()?;

        Ok(response)
    }
}

pub fn get_owner_repo_from_url(url: &str) -> Result<(String, String)> {
    let regex = Regex::new(r"http(s)://github\.com/(?P<owner>[^/]+)/(?P<repo>[^/]+)").unwrap();
    let group = regex
        .captures(url)
        .context("Not a valid github repository url")?;
    let owner: String = group.name("owner").unwrap().as_str().to_string();
    let repo: String = group.name("repo").unwrap().as_str().to_string();
    Ok((owner, repo))
}

#[cfg(test)]
mod test {
    use crate::modules::github::get_owner_repo_from_url;

    #[test]
    fn get_owner_repo_from_url_ok() {
        let input = "https://github.com/dibericky/perty";
        let output = get_owner_repo_from_url(input).unwrap();
        assert_eq!(output, ("dibericky".to_owned(), "perty".to_owned()));
    }

    #[test]
    fn get_owner_repo_from_url_missing_repo() {
        let input = "https://github.com/dibericky/";
        let error = get_owner_repo_from_url(input).err().unwrap().to_string();
        assert_eq!(error, "Not a valid github repository url".to_string());
    }

    #[test]
    fn get_owner_repo_from_url_not_github() {
        let input = "https://foobar.com/dibericky/perty";
        let error = get_owner_repo_from_url(input).err().unwrap().to_string();
        assert_eq!(error, "Not a valid github repository url".to_string());
    }
}
