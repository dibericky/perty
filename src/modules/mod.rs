pub mod activity;
pub mod activity_report;
pub mod pert;
pub mod pert_report;
pub mod roadmap;
pub mod roadmap_report;
pub mod storage;
