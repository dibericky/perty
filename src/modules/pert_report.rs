use super::{
    activity::{Activity, EstimationValue},
    pert::Pert,
};
use anyhow::Result;
use cli_table::{format::Justify, Table, WithTitle};
use serde::Serialize;
use std::path::Path;

type Partials = liquid::partials::EagerCompiler<liquid::partials::InMemorySource>;

fn get_partials() -> Result<Partials> {
    let mut partials = Partials::empty();
    let current_dir = std::env::current_dir().unwrap();
    let relative_template_folder = "src/modules/templates/_includes".to_string();
    let partials_folder_path = current_dir.join(relative_template_folder);
    let folder_str = partials_folder_path.to_str().unwrap();
    let paths = std::fs::read_dir(&folder_str).unwrap();
    for entry in paths {
        let entry = entry.unwrap();
        let file_name = entry.file_name();
        let file_name = file_name.to_str().unwrap();
        let file_path = entry.path();
        let file_path = file_path.to_str().unwrap();
        let content = std::fs::read_to_string(file_path).expect("failed to read partial file");
        partials.add(file_name, content);
    }

    Ok(partials)
}

#[derive(Table, Serialize)]
pub struct PertReportRow {
    #[table(title = "Activity", justify = "Justify::Right")]
    name: String,
    #[table(title = "Optimistic")]
    optimistic: EstimationValue,
    #[table(title = "Probable")]
    probable: EstimationValue,
    #[table(title = "Pessimistic")]
    pessimistic: EstimationValue,
    #[table(title = "PERT estimation")]
    pert: f64,
}

pub struct PertReport {
    pert: Pert,
    rows: Vec<PertReportRow>,
    estimated_total: f64,
}

impl From<&Activity> for PertReportRow {
    fn from(act: &Activity) -> PertReportRow {
        Self {
            name: act.name.to_owned(),
            optimistic: act.estimation.optimistic,
            probable: act.estimation.probable,
            pessimistic: act.estimation.pessimistic,
            pert: act.estimation.estimated(),
        }
    }
}

fn estimated_total(activities: &[Activity]) -> f64 {
    activities.iter().map(|activity| activity.estimated()).sum()
}

impl PertReport {
    pub fn new(pert: Pert, activities: Vec<Activity>) -> Self {
        Self {
            pert,
            rows: activities.iter().map(|act| (act).into()).collect(),
            estimated_total: estimated_total(&activities),
        }
    }

    pub fn csv(&mut self) -> String {
        let mut csv_rows = vec![vec![
            "Name".to_string(),
            "Optimistic".to_string(),
            "Probable".to_string(),
            "Pessimistic".to_string(),
            "PERT".to_string(),
        ]];

        let mut body_rows = self
            .rows
            .iter()
            .map(|row| {
                vec![
                    row.name.to_owned(),
                    row.optimistic.to_string(),
                    row.probable.to_string(),
                    row.pessimistic.to_string(),
                    row.pert.to_string(),
                ]
            })
            .collect::<Vec<_>>();

        csv_rows.append(&mut body_rows);
        csv_rows.push(vec![
            "Total".to_string(),
            "".to_string(),
            "".to_string(),
            "".to_string(),
            self.estimated_total.to_string(),
        ]);

        csv_rows
            .into_iter()
            .map(|vec_row| vec_row.join(","))
            .collect::<Vec<_>>()
            .join("\n")
    }

    pub fn html(&mut self) -> String {
        let partials = get_partials().unwrap();

        let path = Path::new("src/modules/templates/report_pert.liquid");
        let template = liquid::ParserBuilder::with_stdlib()
            .partials(partials)
            .build()
            .unwrap();

        let template = template.parse_file(path).unwrap();

        let globals = liquid::object!({
            "pert_name": self.pert.name,
            "activities": self.rows,
            "estimated_total": self.estimated_total
        });

        template.render(&globals).unwrap()
    }

    pub fn ascii(&mut self) -> String {
        let rows_str = self.rows.with_title().display().unwrap().to_string();

        format!(
            "Project: {}\n\n{}\nTOTAL: {}",
            self.pert.name, rows_str, self.estimated_total
        )
    }
}

#[cfg(test)]
mod test {
    use super::PertReport;
    use crate::modules::{activity::Activity, pert::Pert};

    #[test]
    fn table() {
        let pert = Pert::new(1, "example".to_string());
        let activities = vec![
            Activity::new(1, "activity 1".to_string(), 6, 10, 15),
            Activity::new(2, "activity 2".to_string(), 18, 25, 39),
            Activity::new(3, "activity 3".to_string(), 14, 22, 35),
            Activity::new(4, "activity 4".to_string(), 23, 34, 62),
        ];

        let mut report = PertReport::new(pert, activities);
        insta::assert_display_snapshot!(report.ascii());
    }

    #[test]
    fn table_html() {
        let pert = Pert::new(1, "example".to_string());
        let activities = vec![
            Activity::new(1, "activity 1".to_string(), 6, 10, 15),
            Activity::new(2, "activity 2".to_string(), 18, 25, 39),
            Activity::new(3, "activity 3".to_string(), 14, 22, 35),
            Activity::new(4, "activity 4".to_string(), 23, 34, 62),
        ];

        let mut report = PertReport::new(pert, activities);
        insta::assert_display_snapshot!(report.html());
    }

    #[test]
    fn table_csv() {
        let pert = Pert::new(1, "example".to_string());
        let activities = vec![
            Activity::new(1, "activity 1".to_string(), 6, 10, 15),
            Activity::new(2, "activity 2".to_string(), 18, 25, 39),
            Activity::new(3, "activity 3".to_string(), 14, 22, 35),
            Activity::new(4, "activity 4".to_string(), 23, 34, 62),
        ];

        let mut report = PertReport::new(pert, activities);
        insta::assert_display_snapshot!(report.csv());
    }
}
