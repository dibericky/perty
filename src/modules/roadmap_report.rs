use cli_table::{format::Justify, Table, WithTitle};
use serde::Serialize;

use super::{
    activity::ActivityId,
    roadmap::{ActivitySum, Phase, Roadmap},
};

#[derive(Table, Serialize)]
struct ActivityRow {
    #[table(title = "ID", justify = "Justify::Right")]
    id: ActivityId,
    #[table(title = "Name")]
    name: String,
}

impl From<ActivitySum> for ActivityRow {
    fn from(act: ActivitySum) -> Self {
        ActivityRow {
            id: act.id,
            name: act.name,
        }
    }
}

struct PhasesRow {
    id: usize,
    activities: Vec<ActivityRow>,
}

impl From<Roadmap> for Vec<PhasesRow> {
    fn from(roadmap: Roadmap) -> Self {
        roadmap
            .phases
            .into_iter()
            .enumerate()
            .map(|(id, ph)| {
                let activities = ph.activities.into_iter().map(|act| act.into()).collect();
                PhasesRow {
                    id: id + 1,
                    activities,
                }
            })
            .collect()
    }
}

pub struct RoadmapReport {
    phases: Vec<PhasesRow>,
}

impl RoadmapReport {
    pub fn new(roadmap: Roadmap) -> Self {
        Self {
            phases: roadmap.into(),
        }
    }

    // pub fn csv(&mut self) -> String {
    //     let mut csv_rows = vec![vec![
    //         "Name".to_string(),
    //         "Optimistic".to_string(),
    //         "Probable".to_string(),
    //         "Pessimistic".to_string(),
    //         "PERT".to_string(),
    //     ]];

    //     let mut body_rows = self
    //         .rows
    //         .iter()
    //         .map(|row| {
    //             vec![
    //                 row.name.to_owned(),
    //                 row.optimistic.to_string(),
    //                 row.probable.to_string(),
    //                 row.pessimistic.to_string(),
    //                 row.pert.to_string(),
    //             ]
    //         })
    //         .collect::<Vec<_>>();

    //     csv_rows.append(&mut body_rows);
    //     csv_rows.push(vec![
    //         "Total".to_string(),
    //         "".to_string(),
    //         "".to_string(),
    //         "".to_string(),
    //         self.estimated_total.to_string(),
    //     ]);

    //     csv_rows
    //         .into_iter()
    //         .map(|vec_row| vec_row.join(","))
    //         .collect::<Vec<_>>()
    //         .join("\n")
    // }

    // pub fn html(&mut self) -> String {
    //     let partials = get_partials().unwrap();

    //     let path = Path::new("src/modules/templates/report_pert.liquid");
    //     let template = liquid::ParserBuilder::with_stdlib()
    //         .partials(partials)
    //         .build()
    //         .unwrap();

    //     let template = template.parse_file(path).unwrap();

    //     let globals = liquid::object!({
    //         "pert_name": self.pert.name,
    //         "activities": self.rows,
    //         "estimated_total": self.estimated_total
    //     });

    //     template.render(&globals).unwrap()
    // }

    pub fn ascii(&mut self) -> String {
        self.phases
            .iter()
            .map(|phase| {
                format!(
                    "Phase #{}\n{}",
                    phase.id,
                    phase.activities.with_title().display().unwrap().to_string()
                )
            })
            .collect::<Vec<String>>()
            .join("\n\n")
    }
}
