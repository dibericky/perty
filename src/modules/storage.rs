use anyhow::Result;
use postgres::{Client, NoTls};

use super::{
    activity::Activity,
    pert::{Pert, PertId},
};

pub trait Storage {
    fn add_pert(&mut self, pert_id: PertId, name: &str) -> Result<()>;
    fn get_pert(&mut self, pert_id: PertId) -> Result<Option<Pert>>;
    fn add_activity(&mut self, pert_id: PertId, activity: Activity) -> Result<()>;
    fn get_activities(&mut self, pert_id: PertId) -> Result<Vec<Activity>>;
}

pub struct PostgresDb {
    client: Client,
}

impl PostgresDb {
    pub fn new() -> Result<Self> {
        let conn_str = std::env::var("POSTGRES_CONN_STR")?;
        let client = Client::connect(&conn_str, NoTls)?;
        Ok(Self { client })
    }
}

impl Storage for PostgresDb {
    fn add_activity(&mut self, pert_id: PertId, activity: Activity) -> Result<()> {
        self.client.execute(
            "INSERT INTO activities (pert_id, pessimistic, probable, optimistic) VALUES ($1, $2, $3, $4)",
            &[
                &pert_id,
                &activity.estimation.pessimistic,
                &activity.estimation.probable, 
                &activity.estimation.optimistic, 
            ],
        )?;

        Ok(())
    }

    fn get_activities(&mut self, pert_id: PertId) -> Result<Vec<Activity>> {
        let rows = self.client.query(
            "
    SELECT
        pert.name as pert_name,
        pert_id,
        act.id as activity_id,
        pessimistic, probable, optimistic,
        act.name as activity_name
    FROM activities as act
    inner join pert on act.pert_id = pert.id
    WHERE pert_id = $1
	",
            &[&pert_id],
        )?;
        let activities = rows
            .into_iter()
            .map(|row| {
                Activity::new(
                    row.get("activity_name"),
                    row.get("optimistic"),
                    row.get("probable"),
                    row.get("pessimistic"),
                )
            })
            .collect::<Vec<_>>();
        Ok(activities)
    }

    fn add_pert(&mut self, pert_id: PertId, name: &str) -> Result<()> {
        self.client.execute(
            "INSERT INTO pert (id, name) VALUES ($1, $2)",
            &[&pert_id, &name],
        )?;

        Ok(())
    }

    fn get_pert(&mut self, pert_id: PertId) -> Result<Option<Pert>> {
        let res = match self
            .client
            .query_one("SELECT * FROM pert WHERE id = $1", &[&pert_id])
        {
            Err(_) => None,
            Ok(row) => Some(Pert::new(pert_id, row.get("name"))),
        };
        Ok(res)
    }
}
