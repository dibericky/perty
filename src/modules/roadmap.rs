use super::activity::ActivityId;

#[derive(Debug)]
pub struct ActivityWithRelatedDependencies {
    pub activity_id: ActivityId,
    pub head_name: String,
    pub activity_id_head: Option<ActivityId>,
}

#[derive(Debug, PartialEq)]
pub struct ActivitySum {
    pub id: ActivityId,
    pub name: String,
}
#[derive(Debug, PartialEq)]
pub struct Phase {
    pub activities: Vec<ActivitySum>,
}

#[derive(Default)]
pub struct Roadmap {
    pub phases: Vec<Phase>,
}

impl Roadmap {
    pub fn new(acts_with_deps: Vec<ActivityWithRelatedDependencies>) -> Self {
        Self {
            phases: get_phases(acts_with_deps),
        }
    }
}

fn get_phases_recursive(
    acts_with_deps: Vec<ActivityWithRelatedDependencies>,
    previous_ids: Option<&Vec<ActivityId>>,
    head_vec: &mut Vec<Phase>,
) {
    if acts_with_deps.is_empty() {
        return;
    }
    let (heads, tails): (
        Vec<Option<ActivitySum>>,
        Vec<Option<ActivityWithRelatedDependencies>>,
    ) = acts_with_deps
        .into_iter()
        .map(|act| match act.activity_id_head {
            None => (
                Some(ActivitySum {
                    id: act.activity_id,
                    name: act.head_name,
                }),
                None,
            ),
            Some(head_id) => {
                if previous_ids.is_some() && previous_ids.unwrap().contains(&head_id) {
                    return (
                        Some(ActivitySum {
                            id: act.activity_id,
                            name: act.head_name,
                        }),
                        None,
                    );
                }
                (None, Some(act))
            }
        })
        .unzip();

    let heads: Vec<ActivitySum> = heads.into_iter().flatten().collect();
    let tails: Vec<ActivityWithRelatedDependencies> = tails.into_iter().flatten().collect();

    let head_ids: Vec<ActivityId> = heads.iter().map(|act| act.id).collect();

    head_vec.push(Phase { activities: heads });

    get_phases_recursive(tails, Some(&head_ids), head_vec)
}

fn get_phases(acts_with_deps: Vec<ActivityWithRelatedDependencies>) -> Vec<Phase> {
    let mut accumulator: Vec<Phase> = vec![];
    get_phases_recursive(acts_with_deps, None, &mut accumulator);

    accumulator
}

#[cfg(test)]
mod test {
    use crate::modules::roadmap::ActivityWithRelatedDependencies;

    use super::{get_phases, ActivitySum, Phase};

    #[test]
    fn test_get_phases_empty() {
        let acts_with_deps: Vec<ActivityWithRelatedDependencies> = vec![];
        let roadmap = get_phases(acts_with_deps);
        let expected: Vec<Phase> = vec![];
        assert_eq!(roadmap, expected)
    }

    #[test]
    fn test_get_phases_only_first_phase() {
        let acts_with_deps: Vec<ActivityWithRelatedDependencies> = vec![
            ActivityWithRelatedDependencies {
                head_name: "First A".to_string(),
                activity_id: 1,
                activity_id_head: None,
            },
            ActivityWithRelatedDependencies {
                head_name: "First B".to_string(),
                activity_id: 2,
                activity_id_head: None,
            },
        ];
        let roadmap = get_phases(acts_with_deps);
        let expected: Vec<Phase> = vec![Phase {
            activities: vec![
                ActivitySum {
                    name: "First A".to_string(),
                    id: 1,
                },
                ActivitySum {
                    name: "First B".to_string(),
                    id: 2,
                },
            ],
        }];
        assert_eq!(roadmap, expected)
    }

    #[test]
    fn test_get_phases_more_phase() {
        let acts_with_deps: Vec<ActivityWithRelatedDependencies> = vec![
            ActivityWithRelatedDependencies {
                head_name: "First A".to_string(),
                activity_id: 1,
                activity_id_head: None,
            },
            ActivityWithRelatedDependencies {
                head_name: "Second B".to_string(),
                activity_id: 2,
                activity_id_head: Some(1),
            },
            ActivityWithRelatedDependencies {
                head_name: "Third C".to_string(),
                activity_id: 3,
                activity_id_head: Some(2),
            },
            ActivityWithRelatedDependencies {
                head_name: "Second D".to_string(),
                activity_id: 4,
                activity_id_head: Some(1),
            },
        ];
        let roadmap = get_phases(acts_with_deps);
        let expected: Vec<Phase> = vec![
            Phase {
                activities: vec![ActivitySum {
                    name: "First A".to_string(),
                    id: 1,
                }],
            },
            Phase {
                activities: vec![
                    ActivitySum {
                        name: "Second B".to_string(),
                        id: 2,
                    },
                    ActivitySum {
                        name: "Second D".to_string(),
                        id: 4,
                    },
                ],
            },
            Phase {
                activities: vec![ActivitySum {
                    name: "Third C".to_string(),
                    id: 3,
                }],
            },
        ];
        assert_eq!(roadmap, expected)
    }
}
