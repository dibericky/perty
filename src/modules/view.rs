use super::{
    activity::{Activity, EstimationValue},
    pert::{Pert, PertId},
};
use cli_table::{format::Justify, Table, WithTitle};

#[derive(Table)]
struct Row {
    #[table(title = "Activity", justify = "Justify::Right")]
    name: String,
    #[table(title = "Optimistic")]
    optimistic: EstimationValue,
    #[table(title = "Probable")]
    probable: EstimationValue,
    #[table(title = "Pessimistic")]
    pessimistic: EstimationValue,
    #[table(title = "PERT estimation")]
    pert: f64,
}

struct PertWithActivities {
    pert: Pert,
    activities: Vec<Activity>,
}
pub struct Report {
    data: PertWithActivities,
}

impl Report {
    pub fn new(pert: Pert, activities: Vec<Activity>) -> Self {
        Self {
            data: PertWithActivities { pert, activities },
        }
    }

    fn estimated_total(&self) -> f64 {
        self.data
            .activities
            .iter()
            .map(|activity| activity.estimated())
            .sum()
    }

    pub fn table(&mut self) -> String {
        let rows: Vec<Row> = self
            .data
            .activities
            .iter()
            .map(|activity| Row {
                name: activity.name.to_owned(),
                pessimistic: activity.estimation.pessimistic,
                probable: activity.estimation.probable,
                optimistic: activity.estimation.optimistic,
                pert: activity.estimated(),
            })
            .collect();
        let rows_str = rows.with_title().display().unwrap().to_string();

        format!(
            "Project: {}\n\n{}\nTOTAL: {}",
            self.data.pert.name,
            rows_str,
            self.estimated_total()
        )
    }
}

#[derive(Table)]
struct ListRowPert {
    #[table(title = "ID", justify = "Justify::Right")]
    id: PertId,
    #[table(title = "Name")]
    name: String,
}

pub fn list_view(perts: Vec<Pert>) -> String {
    let list: Vec<ListRowPert> = perts
        .into_iter()
        .map(|pert| ListRowPert {
            id: pert.id,
            name: pert.name,
        })
        .collect();
    list.with_title().display().unwrap().to_string()
}

#[cfg(test)]
mod test {
    use super::Report;
    use crate::modules::{activity::Activity, pert::Pert};

    #[test]
    fn table() {
        let pert = Pert::new(1, "example".to_string());
        let activities = vec![
            Activity::new("activity 1".to_string(), 6, 10, 15),
            Activity::new("activity 2".to_string(), 18, 25, 39),
            Activity::new("activity 3".to_string(), 14, 22, 35),
            Activity::new("activity 4".to_string(), 23, 34, 62),
        ];

        let mut report = Report::new(pert, activities);
        insta::assert_display_snapshot!(report.table());
    }
}
