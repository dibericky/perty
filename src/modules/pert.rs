use anyhow::Result;

use super::{activity::Activity, storage::Storage};

pub struct Pert {
    pub name: String,
    pub activities: Box<dyn Storage>,
}

impl Pert {
    pub fn new(name: String, storage: Box<dyn Storage>) -> Self {
        Self {
            name,
            activities: storage,
        }
    }

    pub fn add_activity(&mut self, activity: Activity) -> &mut Self {
        self.activities.add_activity(&self.name, activity).unwrap();
        self
    }

    pub fn get_activities (&mut self) -> Result<Vec<&Activity>> {
        self.activities
            .get_activities(&self.name)
    }

    pub fn estimated_total(&mut self) -> f64 {
        self
            .get_activities()
            .unwrap()
            .iter()
            .map(|activity| activity.estimated())
            .sum()
    }
}

#[cfg(test)]
mod test {
    use crate::modules::storage::{MemoryStorage};

    use super::*;

    #[test]
    fn add_activity() {
        let mut storage = MemoryStorage::new();
        let activity0 = Activity::new("activity 0".to_string(), 0, 10, 30);
        storage.add_activity("example", activity0).unwrap();

        let mut pert = Pert::new("example".to_string(), Box::new(storage));
        // let activity
        let activity = Activity::new("activity 1".to_string(), 10, 20, 30);
        pert.add_activity(activity);

        let activity0 = Activity::new("activity 0".to_string(), 0, 10, 30);
        let activity = Activity::new("activity 1".to_string(), 10, 20, 30);
        let expected = vec![
            &activity0,
            &activity
        ];

        assert_eq!(pert.name, "example".to_string());
        assert_eq!(pert.get_activities().unwrap(), expected);
    }

    #[test]
    fn estimated_total() {
        let storage = MemoryStorage::new();
        let mut pert = Pert::new("example".to_string(), Box::new(storage));
        pert.add_activity(Activity::new("activity 1".to_string(), 6, 10, 15))
            .add_activity(Activity::new("activity 2".to_string(), 18, 25, 39))
            .add_activity(Activity::new("activity 3".to_string(), 14, 22, 35))
            .add_activity(Activity::new("activity 4".to_string(), 23, 34, 62));

        assert_eq!(pert.estimated_total(), 95.99999999999999f64);
    }
}
