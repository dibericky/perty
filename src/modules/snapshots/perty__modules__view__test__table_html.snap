---
source: src/modules/view.rs
expression: report.table_html()
---
<html>
    <head>
        <style>
            *{
                box-sizing: border-box;
                -webkit-box-sizing: border-box;
                -moz-box-sizing: border-box;
            }
            body{
                font-family: Helvetica;
                -webkit-font-smoothing: antialiased;
                background: rgba( 71, 147, 227, 1);
            }
            h2{
                text-align: center;
                font-size: 18px;
                text-transform: uppercase;
                letter-spacing: 1px;
                color: white;
                padding: 30px 0;
            }

            /* Table Styles */

            .pert-wrapper {
                margin: 10px 70px 70px;
                display: flex;
                flex-direction: column;
                text-align: center;
            }

            table {
                box-shadow: 0px 35px 50px rgba( 0, 0, 0, 0.2 );
                border-radius: 5px;
                font-size: 12px;
                font-weight: normal;
                border: none;
                border-collapse: collapse;
                /* width: 100%; */
                max-width: 100%;
                white-space: nowrap;
                background-color: white;
            }

            table td, table th {
                text-align: center;
                padding: 8px;
            }

            table td {
                font-size: 12px;
            }

            table thead th {
                color: #ffffff;
                background: #4FC3A1;
            }


            table thead th:nth-child(odd) {
                color: #ffffff;
                background: #324960;
            }

            table tr:nth-child(even) {
                background: #F8F8F8;
            }

            table tr:last-child td {
                font-weight: bold
            }
            table tr:last-child, table tr:last-child td:first-child, table tr:last-child td:last-child {
                border-radius: 0px 0px 5px 5px;
            }
        </style>
    </head>
    <body>
        <div class="pert-wrapper">
            <h2>example</h2>
            <table>
    <tr>
        <th>Name</th>
        <th>Optimistic</th>
        <th>Probable</th>
        <th>Pessimistic</th>
        <th>Pert</th>
    </tr>
    
    <tr>
        <td>activity 1</td>
        <td>6</td>
        <td>10</td>
        <td>15</td>
        <td>10.166666666666666</td>
    </tr>
    
    <tr>
        <td>activity 2</td>
        <td>18</td>
        <td>25</td>
        <td>39</td>
        <td>26.166666666666664</td>
    </tr>
    
    <tr>
        <td>activity 3</td>
        <td>14</td>
        <td>22</td>
        <td>35</td>
        <td>22.833333333333332</td>
    </tr>
    
    <tr>
        <td>activity 4</td>
        <td>23</td>
        <td>34</td>
        <td>62</td>
        <td>36.83333333333333</td>
    </tr>
    
    <td colspan="2">
        Total
    </td>
    <td colspan="3">
        95.99999999999999
    </td>
</table>
        </div>
    </body>
</html>
